### PR TITLE
Initial version of 2.0 version

### PR DESCRIPTION
**Pull request by joelwurtz**
This is a preview of the 2.0 version if some people wants to give a try, it should be already functional, however there are few things that needs to be done before merging / releasing.

What's planned

 * This library will now follow semver and support multiple version of the docker api by using different namespaces
 * Generated code comes from the official swagger file from docker, however there is some glitches and bad specficiation that needs to be fixed (but it only consist of some bad new lines in the yaml and wrong nullable possibility)
 * It's fully BC break and unfortunately it's too much work to add a compatibility layer :/, unless someone wants to do it (as i use a different class name / functions / namespace, it can be added)
 
Basically you will be able to do the following : 

```
$docker = Docker::create(Docker::VERSION_1_32);
$docker->image->imageBuild(....)
```

What's missing for a release (contributions welcome)

 * [x] More tests
 * [x] Better test system (to test different version of the api on different version of docker)
 * [x] Rewriting docs
 * [x] Better code api docs (not sure if possible)
 * [x] Full Async support with Amp / Artax
 * [x] Symfony / PHP-HTTP Factory Client support: done

### CODE REVIEW

**Comment by Soullivaneuh** [comment]
3400 modified files! :o

Are you just moving 2.x to master?

Which constraint should we use on composer if we want to try 2.x? Is it currently usable even if not stable?

**Comment by joelwurtz** [comment]
> Are you just moving 2.x to master?

Yes, and accept merge request on the 1.24 if there are good enough

**Comment by mburtscher** [comment]
That looks great :+1: 

I've tried it out in our project and I'm experiencing some troubles with the services API. When I try to list or inspect a swarm service, I'm getting `InvalidArgumentException`s in the `TaskSpecContainerSpecPrivilegesCredentialSpecNormalizer` and `TaskSpecContainerSpecPrivilegesSELinuxContextNormalizer` because `NULL` is passed as `$data` to the `denormalize()` method.

**Comment by boonkerz** [comment]
yea same issue because i think this stuff is optional.
as example the credential is windows only

I have an other issue. I can not filter like:
`$worker = $docker->service()->serviceList(['filters' => ['name' => 'worker']]);`
or:
`$worker = $docker->service()->serviceList(['filters' => ['label' => 'worker']]);`

**Comment by Soullivaneuh** [comment]
@mburtscher @boonkerz You may opened dedicated issue for that.

**Comment by SamMousa** [comment]
@joelwurtz I've provided some added tests in #276, the code coverage is pretty high even though test quality of some tests could be improved. 

Can you be more specific in your goal / requirement before a release?

You mention rewriting docs before a release, is this a hard requirement? 
While I understand that docs are needed, from a usage perspective what developers care about is that the API of the library doesn't have breaking changes.

The sooner the API is "frozen" for the initial 2.0 release the sooner people can start using this and that might increase the contributions.

**Comment by joelwurtz** [comment]
> Can you be more specific in your goal / requirement before a release?

Like any open source project, i mainly start working on docker-php for my own needs, and know my number one requirement is to have an async version of docker-php, however this can be done in a 2.1 release, so not so hard.

> You mention rewriting docs before a release, is this a hard requirement?

Not really, but it must provide docs on how the installation is different from the last version (as you will be able to have a specific version, and also the dependent library is not semver so there should be some warnings)

In fact there is no real hard requirement, but i didn't have the time to test differents API point and i don't really want to create a release where i may break BC because it's not working

**Comment by SamMousa** [comment]
> Not really, but it must provide docs on how the installation is different from the last version (as you will be able to have a specific version, and also the dependent library is not semver so there should be some warnings)

Can you elaborate on what you mean by `will be able to have a specific version`? I can't find references to that in the repository.

**Comment by joelwurtz** [comment]
The 2.0 version will have a requiremend on `docker-php/docker-php-api`: `4.1.*` this requirement define the docker api version, as an example you will be able to use the 1.25 api of docker by forcing the dependency on this library to `4.1.25.*` or the 1.32 by having `4.1.32.*`

**Comment by Soullivaneuh** [comment]
I just migrated to this branch and for now it works perfectly! :+1: 

Any ETA for a release?

**Comment by joelwurtz** [comment]
I'm currently in the process of changing some points in Jane for the 4.0. Where the goal is to have better inheritance support for custom api endpoint (as for now there is a lot to rewrite to have specific behavior / feature in a endpoint)

But i want to release Jane by the end of January and docker php will follow

**Comment by boonkerz** [comment]
actually i have notices:

```
Warning: Declaration of Docker\Resource\ImageResourceTrait::imageBuild($inputStream, array $parameters = Array, string $fetch = self::FETCH_OBJECT) should be compatible with Docker\API\Client::imageBuild(string $inputStream, array $queryParameters = Array, array $headerParameters = Array, string $fetch = self::FETCH_OBJECT) in /home/thomas/projekte/isc/vendor/docker-php/docker-php/src/Docker.php on line 16

Warning: Declaration of Docker\Resource\ImageResourceTrait::imageCreate($inputImage, array $parameters = Array, string $fetch = self::FETCH_OBJECT) should be compatible with Docker\API\Client::imageCreate(string $inputImage, array $queryParameters = Array, array $headerParameters = Array, string $fetch = self::FETCH_OBJECT) in /home/thomas/projekte/isc/vendor/docker-php/docker-php/src/Docker.php on line 16

Warning: Declaration of Docker\Resource\ImageResourceTrait::imagePush(string $name, array $parameters = Array, string $fetch = self::FETCH_OBJECT) should be compatible with Docker\API\Client::imagePush(string $name, array $queryParameters = Array, array $headerParameters = Array, string $fetch = self::FETCH_OBJECT) in /home/thomas/projekte/isc/vendor/docker-php/docker-php/src/Docker.php on line 16

Warning: Declaration of Docker\Resource\ExecResourceTrait::execStart(Docker\API\Model\ExecIdStartPostBody $execStartConfig, string $id, array $parameters = Array, string $fetch = self::FETCH_OBJECT) should be compatible with Docker\API\Client::execStart(string $id, Docker\API\Model\ExecIdStartPostBody $execStartConfig, string $fetch = self::FETCH_OBJECT) in /home/thomas/projekte/isc/vendor/docker-php/docker-php/src/Docker.php on line 16
```

**Comment by Soullivaneuh** [comment]
@boonkerz please see https://github.com/docker-php/docker-php/issues/294
